File: Flutter/trellon_flutter/lib/core/constants/card_status_values.dart
import 'package:flutter/material.dart';

class CardStatusValues {
  static const String completed = 'completed';
  static const String toDo = 'to_do';
  static const String dueSoon = 'due_soon';
  static const String overdue = 'overdue';

  static const String dueDateInPastMessage =
      'Ngày hết hạn không được trong quá khứ';
  static const String startDateAfterDueDateMessage =
      'Ngày bắt đầu không được lớn hơn ngày hết hạn';

  static const Map<String, String> _legacyStatusMap = {
    'hoàn thành': completed,
    'hoan thanh': completed,
    'done': completed,
    'completed': completed,
    'complete': completed,
    'to do': toDo,
    'todo': toDo,
    'to_do': toDo,
    'due_soon': dueSoon,
    'duesoon': dueSoon,
    'due soon': dueSoon,
    'sắp hết hạn': dueSoon,
    'sap het han': dueSoon,
    'overdue': overdue,
    'over due': overdue,
    'hết hạn': overdue,
    'het han': overdue,
  };

  static String normalize(String? status) {
    final key = status?.trim().toLowerCase();
    if (key == null || key.isEmpty) return toDo;
    return _legacyStatusMap[key] ?? toDo;
  }

  static bool isCompleted(String? status) => normalize(status) == completed;
  static bool isOverdue(String? status) => normalize(status) == overdue;
  static bool isDueSoon(String? status) => normalize(status) == dueSoon;

  static String calculate(String? currentStatus, DateTime? dueDate) {
    if (isCompleted(currentStatus)) return completed;
    if (dueDate == null) return toDo;

    final now = DateTime.now().toUtc();
    final dueUtc = dueDate.toUtc();
    if (dueUtc.isBefore(now)) return overdue;
    return dueUtc.difference(now) <= const Duration(days: 1) ? dueSoon : toDo;
  }

  static bool isDueDateInPast(DateTime? dueDate) {
    return dueDate != null && dueDate.toUtc().isBefore(DateTime.now().toUtc());
  }

  static String label(String? status) {
    switch (normalize(status)) {
      case completed:
        return 'Hoàn thành';
      case overdue:
        return 'Hết hạn';
      case dueSoon:
        return 'Sắp hết hạn';
      default:
        return 'Phải làm';
    }
  }

  static Color color(String? status, {Color defaultColor = Colors.transparent}) {
    switch (normalize(status)) {
      case overdue:
        return Colors.red;
      case dueSoon:
        return Colors.amber;
      case completed:
        return Colors.green;
      default:
        return defaultColor;
    }
  }
}

File: C#/TodoAppAPI/Constants/CardStatusValues.cs
using TodoAppAPI.Models;

namespace TodoAppAPI.Constants
{
    public static class CardStatusValues
    {
        public const string Completed = "completed";
        public const string ToDo = "to_do";
        public const string DueSoon = "due_soon";
        public const string Overdue = "overdue";
        private const string LegacyDeleted = "Deleted";

        public const string DueDateInPastMessage = "Ngày hết hạn không được trong quá khứ";
        public const string StartDateAfterDueDateMessage = "Ngày bắt đầu không được lớn hơn ngày hết hạn";

        private static readonly Dictionary<string, string> LegacyStatusMap = new(StringComparer.OrdinalIgnoreCase)
        {
            ["Hoàn thành"] = Completed,
            ["Hoan thanh"] = Completed,
            ["hoan_thanh"] = Completed,
            ["Done"] = Completed,
            ["Completed"] = Completed,
            ["Complete"] = Completed,
            ["complete"] = Completed,
            [Completed] = Completed,
            ["To Do"] = ToDo,
            ["ToDo"] = ToDo,
            ["todo"] = ToDo,
            ["To do"] = ToDo,
            ["Todo"] = ToDo,
            ["TODO"] = ToDo,
            [ToDo] = ToDo,
            ["Due_soon"] = DueSoon,
            ["Duesoon"] = DueSoon,
            ["DueSoon"] = DueSoon,
            ["Due Soon"] = DueSoon,
            ["Sắp hết hạn"] = DueSoon,
            ["Sap het han"] = DueSoon,
            [DueSoon] = DueSoon,
            ["Overdue"] = Overdue,
            ["Over Due"] = Overdue,
            ["Hết hạn"] = Overdue,
            ["Het han"] = Overdue,
            [Overdue] = Overdue
        };

        public static string Normalize(string? status)
        {
            var trimmed = status?.Trim();
            if (string.IsNullOrWhiteSpace(trimmed))
                return ToDo;

            return LegacyStatusMap.TryGetValue(trimmed, out var normalized)
                ? normalized
                : ToDo;
        }

        public static bool IsValid(string? status)
        {
            var trimmed = status?.Trim();
            return trimmed == Completed || trimmed == ToDo || trimmed == DueSoon || trimmed == Overdue;
        }

        public static bool IsCompleted(string? status) => Normalize(status) == Completed;
        public static bool IsOverdue(string? status) => Normalize(status) == Overdue;
        public static bool IsDueSoon(string? status) => Normalize(status) == DueSoon;
        public static bool IsLegacyDeleted(string? status) =>
            string.Equals(status?.Trim(), LegacyDeleted, StringComparison.OrdinalIgnoreCase);

        public static string CalculateStatus(string? currentStatus, DateTime? dueDate, DateTime? nowUtc = null)
        {
            if (IsCompleted(currentStatus))
                return Completed;

            return CalculateOpenStatus(dueDate, nowUtc ?? DateTime.UtcNow);
        }

        public static string CalculateOpenStatus(DateTime? dueDate, DateTime nowUtc)
        {
            if (!dueDate.HasValue)
                return ToDo;

            var dueUtc = ToUtc(dueDate.Value);
            if (dueUtc < nowUtc)
                return Overdue;

            return dueUtc <= nowUtc.AddDays(1) ? DueSoon : ToDo;
        }

        public static bool ApplyCalculatedStatus(Card card, DateTime? nowUtc = null)
        {
            var nextStatus = CalculateStatus(card.Status, card.DueDate, nowUtc ?? DateTime.UtcNow);
            if (card.Status == nextStatus)
                return false;

            card.Status = nextStatus;
            return true;
        }

        public static bool IsDueDateInPast(DateTime? dueDate, DateTime? nowUtc = null)
        {
            return dueDate.HasValue && ToUtc(dueDate.Value) < (nowUtc ?? DateTime.UtcNow);
        }

        public static DateTime ToUtc(DateTime value)
        {
            return value.Kind switch
            {
                DateTimeKind.Utc => value,
                DateTimeKind.Local => value.ToUniversalTime(),
                _ => DateTime.SpecifyKind(value, DateTimeKind.Utc)
            };
        }
    }
}
